File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {number[]} inorder
 * @param {number[]} postorder
 * @return {TreeNode}
 */
var buildTree = function (inorder, postorder) {
    function solve(inorder, postorder, postorderIndex, start, end) {
        if (start > end || postorderIndex.val<0) {
            return null
        }
        var postordervalue = postorder[postorderIndex.val];
        let inorderIndex = -1;
        for (let i = start; i <= end; i++) {
            if (inorder[i] == postordervalue) {
                inorderIndex = i;
                break;

            }
        }
        var root = new TreeNode(postordervalue);
        postorderIndex.val = postorderIndex.val - 1;
         root.right = solve(inorder, postorder, postorderIndex, inorderIndex + 1, end)
        root.left = solve(inorder, postorder, postorderIndex, start, inorderIndex - 1)
       
        
        return root
    }


    var n = inorder.length;
    const postorderIndex = {
        val: n - 1
    }

    return solve(inorder, postorder, postorderIndex, 0, n - 1)



};

File: 0106-construct-binary-tree-from-inorder-and-postorder-traversal/0106-construct-binary-tree-from-inorder-and-postorder-traversal.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* helper(vector<int>& inorder, vector<int>& postorder, int n,
                     int low, int high, int &postorderindex) {
        if (low > high) {
            return NULL;
        }
        int inorderIndex = -1;
        for (int i = low; i <= high; i++) {
            if (inorder[i] == postorder[postorderindex]) {
                inorderIndex = i;
                break;
            }
        }
        TreeNode* root = new TreeNode(inorder[inorderIndex]);
        postorderindex--;
        root->right = helper(inorder, postorder, n, inorderIndex + 1, high,
                             postorderindex);
        root->left = helper(inorder, postorder, n, low, inorderIndex - 1,
                            postorderindex);
        
        return root;
    }
    TreeNode* buildTree(vector<int>& inorder, vector<int>& postorder) {

        int n = inorder.size();
        int low = 0;
        int high = n - 1;
        int postorderindex = n - 1;

        return helper(inorder, postorder, n, low, high, postorderindex);
    }
};